### PR TITLE
fix: strip stray <p></p> HTML tags from imported Jupyter markdown cells

### LINKED ISSUE #8651: Jupyter notebook importer leaves stray `<p>` `</p>` HTML tags
### Describe the bug

I imported https://github.com/sagemanifolds/SageManifolds/blob/master/Notebooks/SM_sphere_S2.ipynb using molab.

Various markdown cells in the marimo notebook contained stray HTML tags that prevented the display of marked-up LaTeX formulas.

### Will you submit a PR?

- [ ] Yes

### Environment

molab

### Code to reproduce

_No response_

### PR DESCRIPTION
**Pull request by giulio-leone**
## Summary

When importing Jupyter notebooks (e.g. [SM_sphere_S2.ipynb](https://github.com/sagemanifolds/SageManifolds/blob/master/Notebooks/SM_sphere_S2.ipynb)), markdown cells containing `<p>…</p>` HTML paragraph wrappers are kept verbatim inside `mo.md()`. This breaks LaTeX rendering because the markdown/math renderer does not process LaTeX delimiters that appear inside HTML `<p>` elements.

### Example

**Jupyter markdown cell:**
```html
<p>We declare that $\mathbb{S}^2 = U \cup V$:</p>
```

**Before (broken in marimo):** The `<p>` tags prevent the LaTeX formula from rendering.

**After (this fix):** Converted to plain text:
```
We declare that $\mathbb{S}^2 = U \cup V$:
```

### Fix

Added a `_strip_paragraph_tags()` helper in `marimo/_convert/ipynb/to_ir.py` that removes bare `<p>`/`</p>` tags (including those with attributes like `<p class="...">`) from markdown cell source before passing it to `markdown_to_marimo()`.

Other HTML tags (`<div>`, `<span>`, `<em>`, `<strong>`, etc.) are preserved — only paragraph wrappers are stripped since they are redundant in plain markdown.

Closes #8651

### CODE REVIEW

File: marimo/_convert/ipynb/to_ir.py [hunk 1]
```diff
@@ -39,7 +39,26 @@ class CodeCell:
...
+    """Remove bare ``<p>`` / ``</p>`` HTML tags from markdown source.
+
+    Jupyter markdown cells often wrap content in ``<p>…</p>`` tags which are
+    redundant in plain markdown and can break LaTeX rendering inside
+    ``mo.md()``.  This helper removes them while preserving all other HTML
+    tags and the text content.
+    """
+    return _PARAGRAPH_TAG_RE.sub("", source)
+
+
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply marimo/_convert/ipynb/to_ir.py]: these are good suggestions and we can add test cases for each one

**Comment by giulio-leone** [comment]
Added tests and addressed reviewer feedback:

**Improvements to `_strip_paragraph_tags`:**
- Now skips `<p>`/`</p>` tags inside fenced code blocks (both backtick and tilde fences), so HTML examples in documentation/code are preserved
- Replaces closing `</p>` tags with a newline instead of removing them, preventing adjacent paragraphs from collapsing (e.g. `<p>a</p><p>b</p>` no longer becomes `ab`)
- Cleans up excessive blank lines introduced by replacements

**12 test cases added** in `tests/_convert/ipynb/test_strip_paragraph_tags.py`:
- Basic tag removal
- Tags with attributes (`<p class="...">`)
- Case-insensitive matching
- Adjacent paragraph separation
- Fenced code block preservation (backtick and tilde)
- Multiline paragraphs
- No-tags passthrough
- Empty string
- LaTeX content preservation
- Nested HTML preservation

All tests verified locally.

**Comment by giulio-leone** [comment]
Updated the implementation based on the review feedback:

### Changes
- **Styled `<p>` tags are now preserved** — only bare `<p>` (no attributes) are stripped. Tags like `<p style="color: red">` or `<p class="lead">` remain intact since they carry semantic meaning.
- **Stack-based pair matching** — each `</p>` is matched to its corresponding `<p>` opener. Only pairs where the opener is bare get stripped.
- **Fenced code blocks still protected** — `<p>` tags inside ` ``` ` or `~~~` blocks are never touched.

### Tests added
- `test_preserves_styled_p_tag` — `<p style="color: red">` stays
- `test_preserves_p_with_class` — `<p class="lead">` stays
- `test_preserves_p_with_id` — `<p id="intro">` stays
- `test_mixed_bare_and_styled` — bare stripped, styled preserved in same cell
- `test_adjacent_styled_paragraphs_fully_preserved` — adjacent styled pairs intact
- All existing tests updated and passing

File: marimo/_convert/ipynb/to_ir.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply marimo/_convert/ipynb/to_ir.py]: Thanks @mscolnick! I'll add test cases for these edge cases — specifically:
1. Preserving `<p>` tags inside fenced code blocks
2. Handling adjacent paragraph tags (`<p>a</p><p>b</p>`) with proper separation

Will push an update shortly.

File: marimo/_convert/ipynb/to_ir.py [hunk 2]
```diff
@@ -39,7 +39,86 @@ class CodeCell:
...
 
 
-# Define a type for transforms that add/remove cells
+# Regex patterns for stripping bare <p>/<\/p> tags from Jupyter markdown cells.
+# Only match bare <p> (no attributes) — styled tags like <p style="color: red">
+# carry semantic meaning and must be preserved.
+_BARE_OPEN_P_RE = re.compile(r"<p\s*>", re.IGNORECASE)
+_ANY_OPEN_P_RE = re.compile(r"<p(?:\s[^>]*)?>", re.IGNORECASE)
+_CLOSE_P_RE = re.compile(r"</p>", re.IGNORECASE)
+_FENCED_BLOCK_RE = re.compile(
```
Comment (dmadisetti) [inline marimo/_convert/ipynb/to_ir.py]: I would use from `pymdownx.superfences import RE_NESTED_FENCE_START`

which is a dep of marimo

Comment (giulio-leone) [reply marimo/_convert/ipynb/to_ir.py]: Good suggestion! `RE_NESTED_FENCE_START` from `pymdownx.superfences` would be more robust than a hand-rolled regex since it's already battle-tested and maintained as a marimo dependency. Let me update the implementation to use it.

**Comment by dmadisetti** [comment]
Tests look unrelated. Thanks!
